server/sonar-server/src/main/java/org/sonar/server/authentication/JwtToken.java

```diff
@@ -0,0 +1,190 @@
+ */
+@ServerSide
+public class JwtToken implements Startable {
+
+  private static final Logger LOG = Loggers.get(JwtToken.class);
+
+  private static final String SECRET_KEY_PROPERTY = "sonar.secretKey";
+
+  private static final SignatureAlgorithm SIGNATURE_ALGORITHM = SignatureAlgorithm.HS256;
+
+  private final Settings settings;
+  private final System2 system2;
+  private final UuidFactory uuidFactory;
+
+  private SecretKey secretKey;
+
+  public JwtToken(Settings settings, System2 system2, UuidFactory uuidFactory) {
+    this.settings = settings;
+    this.system2 = system2;
+    this.uuidFactory = uuidFactory;
+  }
+
+  String encode(Jwt jwt) {
+    long now = system2.now();
+    JwtBuilder jwtBuilder = Jwts.builder()
+      .setId(uuidFactory.create())
+      .setSubject(jwt.getUserLogin())
+      .setIssuedAt(new Date(now))
+      .setExpiration(new Date(now + jwt.getExpirationTimeInSeconds() * 1000))
+      .signWith(SIGNATURE_ALGORITHM, secretKey);
+    for (Map.Entry<String, Object> entry : jwt.getProperties().entrySet()) {
+      jwtBuilder.claim(entry.getKey(), entry.getValue());
+    }
+    return jwtBuilder.compact();
+  }
+
+  Optional<Claims> decode(String token) {
+    try {
+      Claims claims = Jwts.parser()
+        .setSigningKey(secretKey)
+        .parseClaimsJws(token)
+        .getBody();
+      checkArgument(claims.getId() != null, "Token id hasn't been found");
+      checkArgument(claims.getSubject() != null, "Token subject hasn't been found");
+      checkArgument(claims.getExpiration() != null, "Token expiration date hasn't been found");
+      checkArgument(claims.getIssuedAt() != null, "Token creation date hasn't been found");
+      return Optional.of(claims);
+    } catch (ExpiredJwtException | SignatureException e) {
+      LOG.trace("Token is expired or secret key has changed", e);
+      return Optional.empty();
+    } catch (Exception e) {
+      throw new InvalidTokenException(e);
+    }
+  }
+
+  @Override
+  public void start() {
+    String encodedKey = settings.getString(SECRET_KEY_PROPERTY);
+    if (encodedKey == null) {
+      SecretKey newSecretKey = generateSecretKey();
+      settings.setProperty(SECRET_KEY_PROPERTY, Base64.getEncoder().encodeToString(newSecretKey.getEncoded()));
+      this.secretKey = newSecretKey;
+    } else {
+      this.secretKey = decodeSecretKey(encodedKey);
+    }
+  }
+
+  private static SecretKey generateSecretKey() {
+    try {
+      return KeyGenerator.getInstance(SIGNATURE_ALGORITHM.getJcaName()).generateKey();
+    } catch (NoSuchAlgorithmException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  private static SecretKey decodeSecretKey(String encodedKey) {
+    byte[] decodedKey = Base64.getDecoder().decode(encodedKey);
+    return new SecretKeySpec(decodedKey, 0, decodedKey.length, SIGNATURE_ALGORITHM.getJcaName());
+  }
+
+  @Override
+  public void stop() {
+    secretKey = null;
+  }
+
+  static class Jwt {
```

Comment:
Ok to rename to JwtSession and remove the builder.
But I think it's to early to merge it into UserSession, it should be done when UserSession is created by the Java WS.